man/portusctl.1.md
PORTUSCTL 1 "portusctl User manuals" "SUSE LLC." "NOVEMBER 2017"
================================================================

# NAME
portusctl \- A client for your Portus instance

# SYNOPSIS

**portusctl** command [command options] [arguments...]

# DESCRIPTION
**portusctl** is a tool that helps administrators perform operation on the
installed Portus instance. For this, administrators will be able to execute
various commands, **configure** the Portus instance, fetch the **logs** produced
by Portus, etc.

# OPTIONS
**portusctl** has some global options that can be set or toggled.

**--server**, **-s**
  The location where the Portus instance serves requests.

**--token**, **-t**
  The authentication token of the user for the Portus REST API.

**--user**, **-u**
  The user of the Portus REST API.

**--quiet**, **-q**
  Prevent portusctl from outputting general information

# COMMANDS
There are two types of commands: those that act upon a resource, and those
others that don't.

The commands that act upon a resource are as follows: **create**, **delete**,
**get**, **update** and **validate**. These commands translate a syntax like
"\<command\> \<resource\> [args...]" into REST API calls, and then present the
given results back to the user. Therefore, you can think of these commands as a
more useful way to target the REST API.

The **exec** command is the only command that doesn't take a resource as its
first argument, and it transforms its arguments into a command that will be run
on the Portus' context. As you can guess, this command assumes that you have
Portus running locally.

# EXAMPLES
Refer to the man page of each command, but here it is a quick example on how to
use **portusctl**:

  $ portusctl get users
  ...(output)...
  ID    Username    Email                Admin    NamespaceID    DisplayName
  1     portus      portus@portus.com    true     2
  2     mssola      admin@portus.test    true     3

  $ portusctl create user username=user email="user@portus.test" password=12341234
  ...(output)...
  ID    Username    Email               Admin    NamespaceID    DisplayName
  3     user        user@portus.test    false    5

  $ portusctl update user 3 display_name=User
  ...(output)...
  ID    Username    Email               Admin    NamespaceID    DisplayName
  3     user        user@portus.test    false    5              User

  $ portusctl delete user 3
  ...(output)...
  Deleted 'user' successfully!

Besides commands dealing with the API itself, you can run arbitrary commands on
the same context as a local Portus instance:

  $ portusctl exec cat .ruby-version
  2.4.2

# BUGS

Please report bugs if you find them at https://github.com/openSUSE/portusctl/issues.

# AUTHOR

Miquel Sabaté Solà \<msabate@suse.com\>

# SEE ALSO

portusctl-exec(1), portusctl-create(1), portusctl-delete(1), portusctl-get(1),
portusctl-update(1) and portusctl-validate(1)

# HISTORY
November 2017, imported from SUSE/Portus by Miquel Sabaté Solà \<msabate@suse.com\>

August 2016, created by Miquel Sabaté Solà \<msabate@suse.com\>


CHANGELOG.md
# Changelog

## Next version

- First release which includes the following features:


portusctl.go
// Copyright (C) 2017 Miquel Sabaté Solà <msabate@suse.com>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package main

import (
	"fmt"

	"gopkg.in/urfave/cli.v1"
)

func main() {
	app := cli.NewApp()
	app.Name = "portusctl"
	app.Usage = "Client for your Portus instance"
	app.UsageText = "portusctl <command> [arguments...]"
	app.HideHelp = true
	app.Version = versionString()

	app.CommandNotFound = func(context *cli.Context, cmd string) {
		fmt.Printf("Incorrect usage: command '%v' does not exist.\n\n", cmd)
		cli.ShowAppHelp(context)
	}

	app.Flags = []cli.Flag{
		cli.StringFlag{
			Name:   "server, s",
			Usage:  "The location where the Portus instance serves requests",
			EnvVar: "PORTUSCTL_API_SERVER",
		},
		cli.StringFlag{
			Name:   "token, t",
			Usage:  "The authentication token of the user for the Portus REST API",
			EnvVar: "PORTUSCTL_API_TOKEN",
		},
		cli.StringFlag{
			Name:   "user, u",
			Usage:  "The user of the Portus REST API",
			EnvVar: "PORTUSCTL_API_USER",
		},
		cli.BoolFlag{
			Name:   "quiet, q",
			Usage:  "Prevent portusctl from outputting general information",
			EnvVar: "PORTUSCTL_QUIET",
		},
	}

	app.Commands = []cli.Command{
		{
			Name:   "create",
			Usage:  "Create the given resource",
			Action: resourceDecorator(create, postAction),
			ArgsUsage: `<resource> [arguments...]

Where <resource> is the resource that you want to create.`,
		},
		{
			Name:   "delete",
			Usage:  "Delete the given resource",
			Action: resourceDecorator(delete, deleteAction),
			ArgsUsage: `<resource> [arguments...]

Where <resource> is the resource that you want to delete.`,
		},
		{
			Name:   "exec",
			Usage:  "Execute an arbitrary command on the environment of your Portus instance",
			Action: execCmd,
			ArgsUsage: `<command> [arguments...]

Where <command> is the command that you want to run on the environment of your
Portus instance. The successive arguments will be passed also to this command.`,
			Flags: []cli.Flag{
				cli.StringFlag{
					Name:   "local, l",
					Value:  defaultPath,
					Usage:  "The location on the current host of your Portus instance",
					EnvVar: "PORTUSCTL_EXEC_LOCATION",
				},
				cli.BoolFlag{
					Name:   "vendor, v",
					Usage:  "Use the local 'vendor' directory as the gem environment",
					EnvVar: "PORTUSCTL_EXEC_VENDOR",
				},
			},
		},
		{
			Name:   "get",
			Usage:  "Fetches info for the given resource",
			Action: resourceDecorator(get, getAction),
			ArgsUsage: `<resource> [arguments...]

Where <resource> is the resource that you want to fetch.`,
			Flags: []cli.Flag{
				cli.StringFlag{
					Name:   "format, f",
					Value:  "default",
					Usage:  "The output format. Available options: default and json",
					EnvVar: "PORTUSCTL_FORMAT",
				},
			},
		},
		{
			Name:      "health",
			Usage:     "Get health info from Portus",
			Action:    healthAction,
			ArgsUsage: " ",
			Flags: []cli.Flag{
				cli.StringFlag{
					Name:   "format, f",
					Value:  "default",
					Usage:  "The output format. Available options: default and json",
					EnvVar: "PORTUSCTL_FORMAT",
				},
			},
		},
		{
			Name:   "update",
			Usage:  "Update the given resource",
			Action: resourceDecorator(update, putAction),
			ArgsUsage: `<resource> [arguments...]

Where <resource> is the resource that you want to update.`,
		},
		{
			Name:   "validate",
			Usage:  "Validate the given resource",
			Action: resourceDecorator(validate, validateAction),
			ArgsUsage: `<resource> [arguments...]

Where <resource> is the resource that you want to validate.`,
			Flags: []cli.Flag{
				cli.StringFlag{
					Name:   "format, f",
					Value:  "default",
					Usage:  "The output format. Available options: default and json",
					EnvVar: "PORTUSCTL_FORMAT",
				},
			},
		},
		{
			Name:      "version",
			Usage:     "Print the client and server version information",
			Action:    versionAction,
			ArgsUsage: " ",
			Flags: []cli.Flag{
				cli.StringFlag{
					Name:   "format, f",
					Value:  "default",
					Usage:  "The output format. Available options: default and json",
					EnvVar: "PORTUSCTL_FORMAT",
				},
			},
		},
	}

	app.RunAndExitOnError()
}
